File: C#/Robot Name/RobotName.cs
using System;
using System.Collections.Generic;

public class Helper
{
    public static char RandomLetter()
    {
        string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

        Random rnd = new Random();
        int index = rnd.Next(0, alphabet.Length);
        return alphabet[index];
    }

    public static int RandomNumber() 
    {
        Random rnd = new Random();
        return rnd.Next(1, 10);
    }
}

public class Robot
{
    protected static HashSet<string> _usedNames = new HashSet<string>();

    string _name = "UNDEFINED";
    public string Name
    {
        get
        {
            return _name;
        }
    }

    private static string GenerateName()
    {
        while (true)
        {
            var name = $"{Helper.RandomLetter()}{Helper.RandomLetter()}{Helper.RandomNumber()}{Helper.RandomNumber()}{Helper.RandomNumber()}";
        
            if(!_usedNames.Contains(name))
            {
                _usedNames.Add(name);
                return name;
            }
        }
    }

    public Robot()
    {
        _name = GenerateName();
    }

    public void Reset()
    {
        _name = GenerateName();
    }
}

File: TypeScript/robot-name.ts
const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const USED_NAMES = new Set();

const randomLetter = (): string => {
  return ALPHABET[Math.floor(Math.random() * 26)];
}

const randomNumber = (): number => {
  return Math.floor(Math.random() * 10);
}

const generateNewName = (): string => {
  while (true) {
    let name = randomLetter() + randomLetter() + String(randomNumber()) + String(randomNumber()) + String(randomNumber());
  
    if(!USED_NAMES.has(name)) {
      USED_NAMES.add(name);
      return name;
    }
  }
}

export class Robot {
  private _name: string;

  constructor() {
      this._name = generateNewName();
  }

  get name() : string{
    return this._name;
  }

  resetName() : void{
    this._name = generateNewName();
  }

  public static releaseNames(): void {
     
  }
}